go/libraries/doltcore/sqle/binlogreplication/binlog_replica_applier.go

```diff
@@ -0,0 +1,997 @@
+	modifiedDatabases map[string]struct{}
+	// currentGtid is the current GTID being processed, but not yet committed
+	currentGtid mysql.GTID
+	// replicationSourceUuid holds the UUID of the source server
+	replicationSourceUuid string
+	// currentPosition records which GTIDs have been successfully executed
+	currentPosition *mysql.Position
+	filters         *filterConfiguration
+}
+
+func newBinlogReplicaApplier(filters *filterConfiguration) *binlogReplicaApplier {
+	return &binlogReplicaApplier{
+		tableMapsById:       make(map[uint64]*mysql.TableMap),
+		stopReplicationChan: make(chan struct{}),
+		modifiedDatabases:   make(map[string]struct{}),
+		filters:             filters,
+	}
+}
+
+// Row Flags – https://mariadb.com/kb/en/rows_event_v1v2-rows_compressed_event_v1/
+
+// rowFlag_endOfStatement indicates that a row event with this flag set is the last event in a statement.
+const rowFlag_endOfStatement = 0x0001
+const rowFlag_noForeignKeyChecks = 0x0002
+const rowFlag_noUniqueKeyChecks = 0x0004
+const rowFlag_noCheckConstraints = 0x0010
+
+// rowFlag_rowsAreComplete indicates that rows in this event are complete, and contain values for all columns of the table.
+const rowFlag_rowsAreComplete = 0x0008
+
+// Go spawns a new goroutine to run the applier's binlog event handler.
+func (a *binlogReplicaApplier) Go(ctx *sql.Context) {
+	go func() {
+		err := a.replicaBinlogEventHandler(ctx)
+		if err != nil {
+			logger.Errorf("unexpected error of type %T: '%v'", err, err.Error())
+			DoltBinlogReplicaController.setSqlError(mysql.ERUnknownError, err.Error())
+		}
+	}()
+}
+
+// connectAndStartReplicationEventStream connects to the configured MySQL replication source, including pausing
+// and retrying if errors are encountered.
+//
+// NOTE: Our fork of Vitess currently only supports mysql_native_password auth. The latest code in the main
+//
+//	Vitess repo supports the current MySQL default auth plugin, caching_sha2_password.
+//	https://dev.mysql.com/blog-archive/upgrading-to-mysql-8-0-default-authentication-plugin-considerations/
+//	To work around this limitation, add the following to your /etc/my.cnf:
+//	    [mysqld]
+//	    default-authentication-plugin=mysql_native_password
+//	or start mysqld with:
+//	    --default-authentication-plugin=mysql_native_password
+func (a *binlogReplicaApplier) connectAndStartReplicationEventStream(ctx *sql.Context) (*mysql.Conn, error) {
+	var maxConnectionAttempts uint64
+	var connectRetryDelay uint32
+	DoltBinlogReplicaController.updateStatus(func(status *binlogreplication.ReplicaStatus) {
+		status.ReplicaIoRunning = binlogreplication.ReplicaIoConnecting
+		status.ReplicaSqlRunning = binlogreplication.ReplicaSqlRunning
+		maxConnectionAttempts = status.SourceRetryCount
+		connectRetryDelay = status.ConnectRetry
+	})
+
+	var conn *mysql.Conn
+	var err error
+	for connectionAttempts := uint64(0); ; connectionAttempts++ {
+		replicaSourceInfo, err := loadReplicationConfiguration(ctx)
```

Comment:
I double-checked the five places that call `loadReplicationConfiguration`. A few need to load it before we apply updates, so those seem reasonable. There is one extra call at the beginning of `StartReplica` that we could _potentially_ remove – it's just there to catch any config validation errors and fail as early as possible. I like that this call on line 128 reloads the configuration here, just in case someone goes in to fix the source connection information while the replica is trying to reconnect. Seems like a nice behavior to refresh that on each connection attempt. Happy to dig in deeper if you see concerns though of course.